### PR TITLE
Feature/streaming jsonrpc batch request

### PR DESCRIPTION
**Pull request by asdacap**
- Stream resuls of jsonrpc request one by one directly to output.
- Significantly reduce memory allocation in case of batched request. 

## Changes

- Refactored JsonRpcResult to contain `IEnumerable` of `JsonRpcResult.Entry`, allowing lazy and streaming processing of batched reuest.

#### What types of changes does your code introduce?

- [X] New feature (a non-breaking change that adds functionality)
#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [X] No: Already have quite a lot of unit tests.

#### Notes on testing

- Tested manually with a python script.

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs [hunk 1]
```diff
@@ -111,42 +107,64 @@ private void IncrementBytesSentMetric(int size)
...
                 _jsonSerializer.Serialize(stream, error);
             }
 
-            await using MemoryStream resultData = new();
-
-            try
+            using (result)
             {
-                if (result.IsCollection)
+                await using MemoryStream resultData = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs:145]: As we are doing such a great job already, maybe we can drop this allocation too? To some reused buffer.

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs:145]: I guess we can.

File: src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs [hunk 2]
```diff
@@ -111,42 +107,64 @@ private void IncrementBytesSentMetric(int size)
...
         public virtual async Task<int> SendJsonRpcResult(JsonRpcResult result)
+        {
+            if (result.IsCollection)
+            {
+                int singleResponseSize = 0;
+                bool isFirst = true;
+                await _handler.SendRawAsync(_jsonOpeningBracket);
+                await foreach (JsonRpcResult.Entry innerResult in result.BatchedResponses!)
+                {
+                    if (!isFirst) await _handler.SendRawAsync(_jsonComma);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs:118]: If we dig deeper, we find:
- In `IpcSocketsHandler`, `_socket.SendAsync(data, SocketFlags.None)`, not sure if we should set `SocketFlags.Truncated` or `SocketFlags.Partial` until we send `_jsonClosingBracket`.
- In `WebSocketHandler` we have `_webSocket.SendAsync(data, WebSocketMessageType.Text, true, CancellationToken.None)`, maybe we should set `endOfMessage` to `false` until we send `_jsonClosingBracket`.

File: src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs:145]: We can do in next PR though

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs:145]: My ideas are to implement either `Stream` or `TextWriter` based on `IByteBuffer`

**Pull request merged**